File: backend/src/xiangqi/xiangqiEngine.ts
import { isGeneralInCheck } from './attackDetector';
import { Board } from './board';
import { BoardPoint } from './boardPoint';
import { decodeFen, encodeFen } from './fenCodec';
import { GameResult, isOngoing, ONGOING, resultForNoMoves, winResult } from './gameResult';
import { allLegalMoves as computeAllLegalMoves, hasAnyLegalMove, legalDestinationsFrom } from './legalMoveGenerator';
import { forMove } from './moveNotation';
import { Move } from './move';
import { opponent, Side } from './side';
import { MoveValidationResult } from './moveValidationResult';
import { pseudoLegalDestinations } from './pieceMoves';

export interface MoveHistoryEntry {
  readonly move: Move;
  readonly notation: string;
  readonly isCheck: boolean;
  readonly resultAfter: GameResult;
}

/**
 * Server-authoritative counterpart of the client's
 * `lib/domain/engine/xiangqi_engine.dart`. This is what
 * `backend/src/rooms/roomManager.ts` uses to validate every multiplayer
 * move - the client's own Dart engine is UX-only for Pass & Play and
 * legal-destination highlighting, never trusted for a live match's result
 * (see docs/ARCHITECTURE.md section 2 and backend/README.md).
 *
 * No undo/redo here (unlike the Dart engine) - a live match has no use for
 * it, and "takebacks" aren't a feature this phase implements.
 */
export class XiangqiEngine {
  private _board: Board;
  private _sideToMove: Side;
  private _result: GameResult;
  private readonly _history: MoveHistoryEntry[] = [];

  constructor(board?: Board, sideToMove: Side = 'red') {
    this._board = board ?? Board.initial();
    this._sideToMove = sideToMove;
    this._result = XiangqiEngine.computeResult(this._board, this._sideToMove);
  }

  static fromFen(fen: string): XiangqiEngine {
    const decoded = decodeFen(fen);
    return new XiangqiEngine(decoded.board, decoded.sideToMove);
  }

  get board(): Board {
    return this._board;
  }

  get sideToMove(): Side {
    return this._sideToMove;
  }

  get result(): GameResult {
    return this._result;
  }

  get history(): ReadonlyArray<MoveHistoryEntry> {
    return this._history;
  }

  isInCheck(side: Side): boolean {
    return isGeneralInCheck(this._board, side);
  }

  get allLegalMoves(): Move[] {
    return isOngoing(this._result) ? computeAllLegalMoves(this._board, this._sideToMove) : [];
  }

  legalDestinationsFrom(from: BoardPoint): BoardPoint[] {
    if (!isOngoing(this._result)) return [];
    return legalDestinationsFrom(this._board, from, this._sideToMove);
  }

  toFen(): string {
    return encodeFen(this._board, this._sideToMove, 0, Math.floor(this._history.length / 2) + 1);
  }

  tryMove(from: BoardPoint, to: BoardPoint): MoveValidationResult {
    if (!isOngoing(this._result)) {
      return { kind: 'illegal', reason: 'gameAlreadyOver', message: 'The game has already ended.' };
    }

    const pc = this._board.pieceAt(from);
    if (pc === null) {
      return { kind: 'illegal', reason: 'noPieceAtSource', message: 'There is no piece on the selected point.' };
    }
    if (pc.side !== this._sideToMove) {
      return { kind: 'illegal', reason: 'notYourTurn', message: "It is not this piece's turn to move." };
    }

    const legal = legalDestinationsFrom(this._board, from, this._sideToMove);
    if (!legal.some((d) => d.row === to.row && d.col === to.col)) {
      const pseudoLegal = pseudoLegalDestinations(this._board, from).some((d) => d.row === to.row && d.col === to.col);
      if (pseudoLegal) {
        return {
          kind: 'illegal',
          reason: 'leavesGeneralInCheck',
          message: 'That move would leave your general in check (or violates the flying-general rule).',
        };
      }
      return { kind: 'illegal', reason: 'illegalPieceMovement', message: 'That piece cannot move there.' };
    }

    const move: Move = { from, to, movedPiece: pc, capturedPiece: this._board.pieceAt(to) };
    const sideBefore = this._sideToMove;

    const nextBoard = this._board.applyMove(from, to);
    const nextSide = opponent(this._sideToMove);
    const opponentInCheck = isGeneralInCheck(nextBoard, nextSide);
    const opponentHasMoves = hasAnyLegalMove(nextBoard, nextSide);
    const isCheckmate = opponentInCheck && !opponentHasMoves;
    const noLegalMoves = !opponentInCheck && !opponentHasMoves;

    const resultAfter = isCheckmate
      ? winResult(sideBefore, 'checkmate')
      : noLegalMoves
        ? winResult(sideBefore, 'noLegalMoves')
        : ONGOING;

    const notation = forMove(move, { isCheck: opponentInCheck, isCheckmate });

    this._history.push({ move, notation, isCheck: opponentInCheck, resultAfter });
    this._board = nextBoard;
    this._sideToMove = nextSide;
    this._result = resultAfter;

    return { kind: 'legal', move, leavesOpponentInCheck: opponentInCheck, endsGame: !isOngoing(resultAfter) };
  }

  /** Ends the game immediately (e.g. clock expiry, resignation) without a move. */
  endGame(winner: Side, reason: 'timeout' | 'resignation'): void {
    this._result = winResult(winner, reason);
  }

  private static computeResult(board: Board, sideToMove: Side): GameResult {
    const inCheck = isGeneralInCheck(board, sideToMove);
    const hasMoves = hasAnyLegalMove(board, sideToMove);
    if (hasMoves) return ONGOING;
    return resultForNoMoves(sideToMove, inCheck);
  }
}

File: lib/domain/engine/xiangqi_engine.dart
import '../fen/fen_codec.dart';
import '../models/board.dart';
import '../models/board_point.dart';
import '../models/chess_clock.dart';
import '../models/game_result.dart';
import '../models/illegal_move_reason.dart';
import '../models/move.dart';
import '../models/move_history_entry.dart';
import '../models/move_validation_result.dart';
import '../models/side.dart';
import '../notation/move_notation.dart';
import '../rules/attack_detector.dart';
import '../rules/legal_move_generator.dart';
import '../rules/piece_moves.dart';

/// The single entry point the presentation layer (and, later, the AI coach
/// and multiplayer sync logic) should depend on. Owns the current board,
/// whose turn it is, move history (for undo/redo and display), the game
/// result, and an optional clock. See RULES_ENGINE.md for the integration
/// contract.
class XiangqiEngine {
  Board _board;
  Side _sideToMove;
  GameResult _result;
  final List<MoveHistoryEntry> _history = [];
  final List<MoveHistoryEntry> _redoStack = [];
  final ChessClock clock;

  XiangqiEngine({
    Board? board,
    Side sideToMove = Side.red,
    ClockConfig clockConfig = const ClockConfig(),
  })  : _board = board ?? Board.initial(),
        _sideToMove = sideToMove,
        _result = const GameResult.ongoing(),
        clock = ChessClock(clockConfig) {
    _result = _computeResult(_board, _sideToMove);
  }

  factory XiangqiEngine.fromFen(
    String fen, {
    ClockConfig clockConfig = const ClockConfig(),
  }) {
    final decoded = FenCodec.decode(fen);
    return XiangqiEngine(
      board: decoded.board,
      sideToMove: decoded.sideToMove,
      clockConfig: clockConfig,
    );
  }

  Board get board => _board;
  Side get sideToMove => _sideToMove;
  GameResult get result => _result;
  List<MoveHistoryEntry> get history => List.unmodifiable(_history);
  bool get canUndo => _history.isNotEmpty;
  bool get canRedo => _redoStack.isNotEmpty;

  bool isInCheck(Side side) => AttackDetector.isGeneralInCheck(_board, side);

  List<Move> get allLegalMoves =>
      _result.isOngoing ? LegalMoveGenerator.allLegalMoves(_board, _sideToMove) : const [];

  List<BoardPoint> legalDestinationsFrom(BoardPoint from) {
    if (!_result.isOngoing) return const [];
    return LegalMoveGenerator.legalDestinationsFrom(_board, from, _sideToMove);
  }

  String toFen() => FenCodec.encode(
        _board,
        _sideToMove,
        fullmoveNumber: (_history.length ~/ 2) + 1,
      );

  /// Validates [from] -> [to] against the current position and, if legal,
  /// executes it: updates the board, flips the turn, records history,
  /// advances the clock increment, and recomputes the game result.
  MoveValidationResult tryMove(BoardPoint from, BoardPoint to) {
    if (!_result.isOngoing) {
      return const IllegalMove(IllegalMoveReason.gameAlreadyOver, 'The game has already ended.');
    }

    final piece = _board.pieceAt(from);
    if (piece == null) {
      return const IllegalMove(
          IllegalMoveReason.noPieceAtSource, 'There is no piece on the selected point.');
    }
    if (piece.side != _sideToMove) {
      return const IllegalMove(
          IllegalMoveReason.notYourTurn, "It is not this piece's turn to move.");
    }

    final legalDestinations = LegalMoveGenerator.legalDestinationsFrom(_board, from, _sideToMove);
    if (!legalDestinations.contains(to)) {
      final isPseudoLegal = pseudoLegalDestinations(_board, from).contains(to);
      if (isPseudoLegal) {
        return const IllegalMove(
          IllegalMoveReason.leavesGeneralInCheck,
          'That move would leave your general in check (or violates the flying-general rule).',
        );
      }
      return const IllegalMove(
          IllegalMoveReason.illegalPieceMovement, 'That piece cannot move there.');
    }

    final move = Move(from: from, to: to, movedPiece: piece, capturedPiece: _board.pieceAt(to));
    final boardBefore = _board;
    final sideBefore = _sideToMove;

    final nextBoard = _board.applyMove(from, to);
    final nextSide = _sideToMove.opponent;
    final opponentInCheck = AttackDetector.isGeneralInCheck(nextBoard, nextSide);
    final opponentHasMoves = LegalMoveGenerator.hasAnyLegalMove(nextBoard, nextSide);
    final isCheckmate = opponentInCheck && !opponentHasMoves;
    final noLegalMoves = !opponentInCheck && !opponentHasMoves;

    final resultAfter = isCheckmate
        ? GameResult.win(sideBefore, GameEndReason.checkmate)
        : noLegalMoves
            ? GameResult.win(sideBefore, GameEndReason.noLegalMoves)
            : const GameResult.ongoing();

    final notation = MoveNotation.forMove(move, isCheck: opponentInCheck, isCheckmate: isCheckmate);

    _history.add(MoveHistoryEntry(
      move: move,
      notation: notation,
      isCheck: opponentInCheck,
      resultAfter: resultAfter,
      boardBefore: boardBefore,
      sideToMoveBefore: sideBefore,
    ));
    _redoStack.clear();

    _board = nextBoard;
    _sideToMove = nextSide;
    clock.applyIncrement(sideBefore);
    _result = resultAfter;

    return LegalMove(move, leavesOpponentInCheck: opponentInCheck, endsGame: !resultAfter.isOngoing);
  }

  bool undo() {
    if (_history.isEmpty) return false;
    final entry = _history.removeLast();
    _redoStack.add(entry);
    _board = entry.boardBefore;
    _sideToMove = entry.sideToMoveBefore;
    _result = const GameResult.ongoing();
    return true;
  }

  bool redo() {
    if (_redoStack.isEmpty) return false;
    final entry = _redoStack.removeLast();
    _history.add(entry);
    _board = entry.boardBefore.applyMove(entry.move.from, entry.move.to);
    _sideToMove = entry.sideToMoveBefore.opponent;
    _result = entry.resultAfter;
    return true;
  }

  static GameResult _computeResult(Board board, Side sideToMove) {
    final inCheck = AttackDetector.isGeneralInCheck(board, sideToMove);
    final hasMoves = LegalMoveGenerator.hasAnyLegalMove(board, sideToMove);
    if (hasMoves) return const GameResult.ongoing();
    return GameResult.win(
        sideToMove.opponent, inCheck ? GameEndReason.checkmate : GameEndReason.noLegalMoves);
  }
}
